packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java

```diff
@@ -0,0 +1,225 @@
+  private void removeClusterManager(Object clusterManagerId) {
+    final ClusterManager<MarkerBuilder> clusterManager =
+        clusterManagerIdToManager.remove(clusterManagerId);
+    if (clusterManager == null) {
+      return;
+    }
+    initListenersForClusterManager(clusterManager, null, null);
+    clusterManager.clearItems();
+    clusterManager.cluster();
+  }
+
+  /** Adds item to the ClusterManager it belongs. */
+  public void addItem(MarkerBuilder item) {
+    ClusterManager<MarkerBuilder> clusterManager =
+        clusterManagerIdToManager.get(item.clusterManagerId());
+    if (clusterManager != null) {
+      clusterManager.addItem(item);
+      clusterManager.cluster();
+    }
+  }
+
+  /** Removes item from the ClusterManager it belongs. */
+  public void removeItem(MarkerBuilder item) {
+    ClusterManager<MarkerBuilder> clusterManager =
+        clusterManagerIdToManager.get(item.clusterManagerId());
+    if (clusterManager != null) {
+      clusterManager.removeItem(item);
+      clusterManager.cluster();
+    }
+  }
+
+  /** Called when ClusterRenderer has rendered new visible marker to the map. */
+  void onClusterItemMarker(MarkerBuilder item, Marker marker) {
+    if (clusterItemListener != null) {
+      clusterItemListener.onClusterItemMarker(item, marker);
+    }
+  }
+
+  /** Reads clusterManagerId from object data. */
+  @SuppressWarnings("unchecked")
+  private static String getClusterManagerId(Object clusterManagerData) {
+    Map<String, Object> clusterMap = (Map<String, Object>) clusterManagerData;
+    return (String) clusterMap.get("clusterManagerId");
+  }
+
+  /**
+   * Requests all current clusters from the algorithm of the requested ClusterManager and converts
+   * them to result response.
+   */
+  public void getClustersWithClusterManagerId(
+      String clusterManagerId, MethodChannel.Result result) {
+    ClusterManager<MarkerBuilder> clusterManager = clusterManagerIdToManager.get(clusterManagerId);
+    if (clusterManager == null) {
+      result.error(
+          "Invalid clusterManagerId", "getClusters called with invalid clusterManagerId", null);
+      return;
+    }
+
+    final Set<? extends Cluster<MarkerBuilder>> clusters =
+        clusterManager.getAlgorithm().getClusters(googleMap.getCameraPosition().zoom);
+    result.success(Convert.clustersToJson(clusterManagerId, clusters));
+  }
+
+  @Override
+  public void onCameraIdle() {
+    for (Map.Entry<String, ClusterManager<MarkerBuilder>> entry :
+        clusterManagerIdToManager.entrySet()) {
+      entry.getValue().onCameraIdle();
+    }
+  }
+
+  @Override
+  public boolean onClusterClick(Cluster<MarkerBuilder> cluster) {
+    if (cluster.getSize() > 0) {
+      MarkerBuilder[] builders = cluster.getItems().toArray(new MarkerBuilder[0]);
+      String clusterManagerId = builders[0].clusterManagerId();
+      methodChannel.invokeMethod("cluster#onTap", Convert.clusterToJson(clusterManagerId, cluster));
```

Comment:
This is fundamental to method channels; it's why we're planning on converting all of our plugins to Pigeon. https://github.com/flutter/flutter/issues/117907 is the tracking issue for this plugin; we're not doing partial migrations since that's more confusing than migrating all at once.